### PR TITLE
Pep8 formatting on astrokat

### PR DESCRIPTION
**Pull request by bngcebetsha**
Initial full run of flake8 formatting on astrokat. Will need to add docstrings to all the functions/methods.

[MT-946](https://skaafrica.atlassian.net/browse/MT-946)

### CODE REVIEW

File: astrokat/test/test_offline_observe.py [hunk 1]
```diff
@@ -36,18 +40,25 @@ def test_two_calib_sim(self):
         cal1 = result.count("0408-65 observed for 30.0 sec")
         cal2 = result.count("1934-638 observed for 30.0 sec")
 
-        self.assertGreaterEqual(cal1 + cal2, 1, "At least one bpcal was observed")
+        self.assertGreaterEqual(
+            cal1 + cal2, 1, "At least one bpcal was observed")
         self.assertLessEqual(cal1 + cal2, 2, "At most 2 bpcals were observed")
```
Comment (bngcebetsha) [inline astrokat/test/test_offline_observe.py]: To fix this error I would need to change the implementation of the test.
`"/home/kat/workspace/946_linting_with_flake8_astrokat/astrokat/test/test_offline_observe.py", line 45, in test_two_calib_sim`

`self.assertLessEqual(cal1 + cal2, 2, "At most 2 bpcals were observed"
AssertionError: 12 not less than or equal to 2 : At most 2 bpcals were observed
"""Fail immediately, with the given message."""
[2019-09-05T10:51:04.076Z] >>  raise self.failureException('12 not less than or equal to 2 : At most 2 bpcals were observed')`

File: astrokat/test/test_offline_observe.py [hunk 2]
```diff
@@ -36,18 +40,25 @@ def test_two_calib_sim(self):
...
+            "Single run through observation target list",
+            result,
+            "Single run")
         self.assertIn("0408-65 observed for 180.0 sec", result)
         self.assertIn("1934-638 observed for 180.0 sec", result)
-        self.assertEqual(result.count("Drift_scan observation for"), 2, "two drift scans")
+        self.assertEqual(
+            result.count("Drift_scan observation for"),
+            2,
+            "two drift scans")
```
Comment (bngcebetsha) [inline astrokat/test/test_offline_observe.py]: Here as well:

`
return func(*args, **keywargs)
File "/home/kat/workspace/946_linting_with_flake8_astrokat/astrokat/test/test_offline_observe.py", line 61, in test_drift_targets_sim
`

`
"two drift scans")
AssertionError: 4 != 2 : two drift scans
'4 != 2' = '%s != %s' % _common_shorten_repr(4, 2)
'4 != 2 : two drift scans' = self._formatMessage('4 != 2 : two drift scans', '4 != 2')
raise self.failureException('4 != 2 : two drift scans')
`

File: astrokat/__init__.py [hunk 3]
```diff
@@ -1,11 +1,12 @@
+"""."""
```
Comment (mmphego) [inline astrokat/__init__.py]: Not sure if this was just a way to get flake8 to pass or not. But a docstring should describe your source code - i.e. what your function, module, or class does. It shouldn't just be a period.

File: astrokat/__main__.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
+"""."""
```
Comment (mmphego) [inline astrokat/__main__.py]: Ditto, if unsure what information to enter. Seek @xinyuwu's input

File: astrokat/__main__.py [hunk 5]
```diff
@@ -81,7 +82,7 @@ def cli(prog,
         short_opts_to_remove=['-h'],
         long_opts_to_remove=['--version'],
         args=None):
-
+    """."""
```
Comment (mmphego) [inline astrokat/__main__.py]: Ditto

File: astrokat/correlator.py [hunk 6]
```diff
@@ -1,9 +1,10 @@
-# Setting correlator configuration values different from default values
+"""Setting correlator configuration values different from default values."""
 
 
 def set_fengines(session,
                  requant_gains=None,
                  fft_shift=None):
+    """."""
```
Comment (mmphego) [inline astrokat/correlator.py]: Ditto

File: astrokat/noisediode.py [hunk 7]
```diff
@@ -1,3 +1,4 @@
+"""."""
```
Comment (mmphego) [inline astrokat/noisediode.py]: Ditto

File: astrokat/noisediode.py [hunk 8]
```diff
@@ -1,3 +1,4 @@
...
 from __future__ import division
 from __future__ import absolute_import
```
Comment (mmphego) [inline astrokat/noisediode.py]: This is ok, but https://stackoverflow.com/a/15011456/6165344

File: astrokat/observatory.py [hunk 9]
```diff
@@ -1,3 +1,4 @@
+"""."""
```
Comment (mmphego) [inline astrokat/observatory.py]: Ditto

File: astrokat/observe_main.py [hunk 10]
```diff
@@ -1,4 +1,5 @@
-# Observation script and chronology check
+"""."""
+# Observation script and chronology check.
```
Comment (mmphego) [inline astrokat/observe_main.py]: Shouldn't this be the docstring?

File: astrokat/observe_main.py [hunk 11]
```diff
@@ -266,40 +272,44 @@ def __exit__(self, type, value, traceback):
...
-        # current sensor list included in instrument
-        # sub_ [
-                # pool_resources,  # ptuse or specific antennas
-                # product,        # correlator product
-                # dump_rate,      # dumprate
-                # band,           # band
-                # ]
+        """Include current sensor list in instrument.
+
+        sub_ [
```
Comment (mmphego) [inline astrokat/observe_main.py]: No need for `#` as this is part of the docstring, from looks of things!

File: scripts/readconfig.py [hunk 12]
```diff
@@ -1,16 +1,21 @@
...
 import yaml
 
 
 def main(yaml_config):
+    """."""
     with open(yaml_config, 'r') as stream:
         try:
             data = yaml.safe_load(stream)
-        except:
+        except BaseException:
```
Comment (mmphego) [inline scripts/readconfig.py]: :+1:

File: astrokat/test/test_offline_observe.py [hunk 13]
```diff
@@ -111,9 +105,10 @@ def test_image_single_sim(self):
         self.assertIn("T4R02C04 observed for 180.0 sec", result)
 
     def test_image_sim(self):
-        """Get result and make sure everything ran properly."""
+        """Test image sim."""
```
Comment (mmphego) [inline astrokat/test/test_offline_observe.py]: Why did you change the docstring, rather update it and give more info.

File: astrokat/test/test_offline_observe.py [hunk 14]
```diff
@@ -65,37 +66,30 @@ def test_drift_targets_sim(self):
...
-    #     self.assertEqual(result.count('Drift_scan observation for'), 2,
-    #    'two drift scans')
+    #     result, 'Single run') self.assertIn('0408-65 observed', result,
+    #     '0408-65 observed') self.assertIn('1934-638 observed', result,
+    #     '1934-638 observed') self.assertEqual(result.count(
+    #     'Drift_scan observation for'), 2, 'two drift scans')
 
     def test_image_single_sim(self):
-        """Get result and make sure everything ran properly."""
+        """Test image single sim."""
```
Comment (mmphego) [inline astrokat/test/test_offline_observe.py]: Less is not more!

**Comment by mmphego** [comment]
Please update `docstrings`, add more information

File: astrokat/__main__.py [hunk 15]
```diff
@@ -88,11 +88,10 @@ def cli(prog,
         # TODO: more complex usage string in separate function
         usage = "%s [options]" \
                 " --yaml <YAMLfile>" % prog
-        description = \
-            "Sources are specified either as part of an observation profile." \
-            " Track one or more sources for a specified time." \
-            " At least one target must be specified." \
-            " Also note the **required** options."
+        description = "Sources are specified either as part of an observation profile." \
```
Comment (mmphego) [inline astrokat/__main__.py]: I read somewhere that using backslashes is to be deprecated sometime in the future, I suggest you handle the long/multiline string differently.

See: https://stackoverflow.com/a/10660443/6165344

File: astrokat/noisediode.py [hunk 16]
```diff
@@ -210,27 +187,27 @@ def pattern(kat,  # kat subarray object
...
-           .format(cycle_length,
-                   float(cycle_length) * float(on_fraction),
-                   nd_antennas))
+    nd_antennas = nd_setup["antennas"]  # selected antennas for nd pattern
+    cycle_length = nd_setup["cycle_len"]  # nd pattern length [sec]
+    on_fraction = nd_setup["on_frac"]  # on fraction of pattern length [%]
+    msg = (
+        "Repeat noise diode pattern every {} sec, with {}"
+        "sec on and apply pattern to {}".format(cycle_length,
+                                                float(cycle_length) * float(on_fraction),
```
Comment (mmphego) [inline astrokat/noisediode.py]: Hanging indentation, perhaps handle it differently.

File: astrokat/observatory.py [hunk 17]
```diff
@@ -89,54 +89,94 @@ def _ephem_settime_(self, ephem_target, lst=True):
         return self.observer.sidereal_time()
 
     def read_file_from_node_config(self, catalogue_file):
-        """Read configuration file."""
+        """Read catalogue file from node config.
+
+        Parameters
+        ----------
+            catalogue_file:
```
Comment (mmphego) [inline astrokat/observatory.py]: Indentation!

See: https://numpydoc.readthedocs.io/en/latest/format.html

File: astrokat/observatory.py [hunk 18]
```diff
@@ -89,54 +89,94 @@ def _ephem_settime_(self, ephem_target, lst=True):
...
         name, target_item = katpoint_target(target_item)
         return self.set_target(target_item)
 
     def unpack_target(self, target_item):
-        """See full description of target."""
+        """Unpack full description of target.
+
+        Parameters
+        -----------
+            target_item: string of target(s)
```
Comment (mmphego) [inline astrokat/observatory.py]: Ditto

File: astrokat/observatory.py [hunk 19]
```diff
@@ -199,18 +243,15 @@ def collect_targets(kat, args):
...
-    msg = ('Found {} target(s): {} from {} catalogue(s), '
-           '{} from default catalogue and {} as target string(s)'
-           .format(len(catalogue),
-                   from_catalogues,
-                   num_catalogues,
-                   from_names,
-                   from_strings))
+    msg = (
+        "Found {} target(s): {} from {} catalogue(s), {} from default"
+        "catalogue and {} as target string(s)".format(len(catalogue), from_catalogues,
```
Comment (mmphego) [inline astrokat/observatory.py]: Hanging indentation, I guess this will be fixed when we run `black`

File: astrokat/scans.py [hunk 20]
```diff
@@ -55,7 +55,7 @@ def raster_scan(session, target, nd_period=None, **kwargs):
 
 
 def scan(session, target, nd_period=None, **kwargs):
-    """Scan basic."""
+    """Scan, basic observation."""
```
Comment (mmphego) [inline astrokat/scans.py]: Not sure about the comma!

File: scripts/astrokat-catalogue2obsfile.py [hunk 21]
```diff
@@ -205,8 +206,12 @@ def read_catalogue(self,
 class BuildObservation(object):
     """Create a default observation config file.
 
-    Assume the format of a target in the list:
-    'name=<name>, radec=<HH:MM:SS.f>,<DD:MM:SS.f>, tags=<tags>, duration=<sec>'
+    Parameters
+    ----------
+        target_list:
```
Comment (mmphego) [inline scripts/astrokat-catalogue2obsfile.py]: Indentation

File: scripts/astrokat-catalogue2obsfile.py [hunk 22]
```diff
@@ -218,14 +223,25 @@ def configure(self,
                   obs_duration=None,
                   lst=None,
                   ):
-        """Configure."""
+        """Set up of the MeerKAT telescope for running observation.
+
+        Parameters
+        ----------
```
Comment (mmphego) [inline scripts/astrokat-catalogue2obsfile.py]: Ditto

**Comment by mmphego** [comment]
As discussed before you merge, I would suggest @ajoubertza to have the final approval.

As for me, all is well.

File: astrokat/observe_main.py [hunk 23]
```diff
@@ -258,217 +263,250 @@ def __enter__(self):
...
+            user_logger.info(
+                "Skipping instrument checks - approved_schedule does not exist"
+            )
             return
         approved_sb_sensor_value = approved_sb_sensor.get_value()
         if self.array.sb_id_code not in approved_sb_sensor_value:
-            user_logger.info('Skipping instrument checks - {} not in approved_schedule'.format(
-                self.array.sb_id_code))
+            user_logger.info(
+                "Skipping instrument checks - {}"
```
Comment (ajoubertza) [inline astrokat/observe_main.py]: There used to be a space after `{}`

File: astrokat/observatory.py [hunk 24]
```diff
@@ -75,61 +80,136 @@ def _ephem_risetime_(self, ephem_target, lst=True):
...
         observer = self.kat.observer
         observer.horizon = numpy.deg2rad(horizon)
         observer.date = ephem.now()
         return observer
 
     def set_target(self, target):
+        """Set the target.
+
+        MeerKAT Wrapper around a PyEphem.Body object,
+        target is
```
Comment (ajoubertza) [inline astrokat/observatory.py]: Sentence incomplete.

File: astrokat/observe_main.py [hunk 25]
```diff
@@ -258,217 +263,250 @@ def __enter__(self):
...
-                # ]
-        user_logger.trace(self.opts.obs_plan_params['instrument'])
-        if self.opts.obs_plan_params['instrument'] is None:
+        """Set up the array for observing.
+
+        Include current sensor list in instrument.
+
+        Parameters
+        ----------
+        instrument: object
```
Comment (ajoubertza) [inline astrokat/observe_main.py]: `object` is very vague.  From the code looks like it is a `dict`, or at least `dict-like`.

File: astrokat/simulate.py [hunk 26]
```diff
@@ -147,74 +183,138 @@ def __exit__(self, type, value, traceback):
...
+
+        How long in seconds, it took to the antennas to move
+        from previous target to the next
+
+        Parameters
+        ----------
+         target: str
+                 a comma-separated description which
+                 contains parameters such as the target name,
+                 position, flux model.
```
Comment (ajoubertza) [inline astrokat/simulate.py]: No.  Implementation is assuming this is `katpoint.Target` object, not `str`.
E.g., usage of `target.body`.

File: astrokat/simulate.py [hunk 27]
```diff
@@ -147,74 +183,138 @@ def __exit__(self, type, value, traceback):
...
+                user_logger.debug("Slewing to {}".format(target.name))
                 slew_time = self.slew_time(target)
         return slew_time
 
     def track(self, target, duration=0, announce=False):
+        """Simulate the track source functionality during observations.
+
+        Parameters
+        ----------
+        target: katpoint.Target object
```
Comment (ajoubertza) [inline astrokat/simulate.py]: When should not include "object" in this, `katpoint.Target` is sufficient - it describes the type.

File: astrokat/noisediode.py [hunk 28]
```diff
@@ -83,42 +76,39 @@ def on(kat,
     else:
         # - use integer second boundary as that is most likely be an exact
         #   time that DMC can execute at
-        msg = ('Dry-run: Set all noise diodes with timestamp {} ({})'
-               .format(np.ceil(timestamp),
-                       time.ctime(timestamp)))
+        msg = "Dry-run: Set all noise diodes with timestamp" "{} ({})".format(
```
Comment (ajoubertza) [inline astrokat/noisediode.py]: The space between `timestamp` and `{}` got lost.

The black auto formatter also results in multiline strings sometimes ending up on one line again, but still in two parts.  This can be fixed as well, but just for this line.

`"Dry-run: Set all noise diodes with timestamp {} ({})".format(`

File: astrokat/noisediode.py [hunk 29]
```diff
@@ -129,108 +119,89 @@ def off(kat,
...
-           'with {} sec on and apply pattern to {}'
-           .format(cycle_length,
-                   float(cycle_length) * float(on_fraction),
-                   nd_antennas))
+    """
+    nd_antennas = nd_setup["antennas"]  # selected antennas for nd pattern
+    cycle_length = nd_setup["cycle_len"]  # nd pattern length [sec]
+    on_fraction = nd_setup["on_frac"]  # on fraction of pattern length [%]
+    msg = (
+        "Repeat noise diode pattern every {} sec, with {} sec on and apply pattern to"
```
Comment (ajoubertza) [inline astrokat/noisediode.py]: Another space that got lost between the last `to` and `{}` on the next line.

File: astrokat/noisediode.py [hunk 30]
```diff
@@ -240,55 +211,55 @@ def pattern(kat,  # kat subarray object
...
-                                                on_fraction,
-                                                cycle_length)
+        replies = kat.ants.req.dig_noise_source(timestamp, on_fraction, cycle_length)
         if not kat.dry_run:
             timestamp = _katcp_reply_to_log_(replies)
         else:
-            msg = ('Dry-run: Set all noise diodes with timestamp {} ({})'
-                   .format(int(timestamp),
-                           time.ctime(timestamp)))
+            msg = "Dry-run: Set all noise diodes with timestamp" "{} ({})".format(
```
Comment (ajoubertza) [inline astrokat/noisediode.py]: Missing space

File: astrokat/observatory.py [hunk 31]
```diff
@@ -191,19 +287,18 @@ def collect_targets(kat, args):
...
         raise ValueError("No known targets found in argument list")
-    msg = ('Found {} target(s): {} from {} catalogue(s), '
-           '{} from default catalogue and {} as target string(s)'
-           .format(len(catalogue),
-                   from_catalogues,
-                   num_catalogues,
-                   from_names,
-                   from_strings))
+    msg = (
+        "Found {} target(s): {} from {} catalogue(s), {} from default catalogue and"
```
Comment (ajoubertza) [inline astrokat/observatory.py]: Missing space at the end of the line

File: astrokat/observe_main.py [hunk 32]
```diff
@@ -62,175 +62,180 @@ def read_targets(target_items):
...
             noisediode.trigger(session.kat, session, duration=nd_period)
-            user_logger.trace('TRACE: ts after nd trigger {}'
-                              .format(time.time()))
-        user_logger.debug('DEBUG: Starting {}s track on target: {} ({})'
-                          .format(duration,
-                                  time.time(),
-                                  time.ctime(time.time())))
+            user_logger.trace("TRACE: ts after nd trigger {}".format(time.time()))
+        user_logger.debug(
+            "DEBUG: Starting {}s track on target:"
```
Comment (ajoubertza) [inline astrokat/observe_main.py]: Missing space

File: astrokat/observe_main.py [hunk 33]
```diff
@@ -258,217 +263,250 @@ def __enter__(self):
...
+                pool_params = [str_.strip() for str_ in conf_param.split(",")]
                 for param in pool_params:
                     if param not in sub_sensor:
-                        raise RuntimeError('Subarray configuration {} error, '
-                                           '{} required, {} found'
-                                           .format(sensor_name,
-                                                   param,
-                                                   sub_sensor))
+                        raise RuntimeError(
+                            "Subarray configuration {} error, {} required,"
```
Comment (ajoubertza) [inline astrokat/observe_main.py]: Missing space

File: astrokat/observe_main.py [hunk 34]
```diff
@@ -258,217 +263,250 @@ def __enter__(self):
...
+                            "{} found".format(sensor_name, param, sub_sensor)
+                        )
             elif conf_param != sub_sensor:
-                raise RuntimeError('Subarray configuration {} error, '
-                                   '{} required, {} found'
-                                   .format(sensor_name,
-                                           conf_param,
-                                           sub_sensor))
+                raise RuntimeError(
+                    "Subarray configuration {} error, {} required,"
```
Comment (ajoubertza) [inline astrokat/observe_main.py]: Missing space

File: astrokat/observe_main.py [hunk 35]
```diff
@@ -515,184 +560,231 @@ def run_observation(opts, kat):
...
-                                user_logger.trace('TRACE: time needed for next obs {} sec'
-                                                  .format(next_cadence_target['cadence']))
-                                next_target = obs_targets[next_cadence_tgt_idx % len(obs_targets)]
+                        for next_cadence_tgt_idx in range(cnt + 1, len(obs_targets)):
+                            next_cadence_target = obs_targets[
+                                next_cadence_tgt_idx % len(obs_targets)
+                            ]
+                            if next_cadence_target["cadence"] > 0:
+                                user_logger.trace(
+                                    "TRACE: time needed for next obs"
```
Comment (ajoubertza) [inline astrokat/observe_main.py]: Missing space

File: astrokat/observe_main.py [hunk 36]
```diff
@@ -515,184 +560,231 @@ def run_observation(opts, kat):
...
         if obs_duration < 0:
-            user_logger.info('Single run through observation target list')
+            user_logger.info("Single run through observation target list")
         else:
-            user_logger.info("Desired observation time {:.2f} sec ({:.2f} min)"
-                             .format(obs_duration, obs_duration/60.))
-        user_logger.info("Total observation time {:.2f} sec ({:.2f} min)"
-                         .format(total_obs_time, total_obs_time/60.))
+            user_logger.info(
+                "Desired observation time {:.2f} sec"
```
Comment (ajoubertza) [inline astrokat/observe_main.py]: Missing space

File: astrokat/observe_main.py [hunk 37]
```diff
@@ -515,184 +560,231 @@ def run_observation(opts, kat):
...
+                "({:.2f} min)".format(obs_duration, obs_duration / 60.0)
+            )
+        user_logger.info(
+            "Total observation time {:.2f} sec"
```
Comment (ajoubertza) [inline astrokat/observe_main.py]: Missing space

File: astrokat/observe_main.py [hunk 38]
```diff
@@ -515,184 +560,231 @@ def run_observation(opts, kat):
...
     # command line arguments will be dropped to unknown arguments
-    unknown_args = [arg for arg in args if arg.startswith('--')]
-    if any('horizon' in arg for arg in unknown_args):
-        raise RuntimeError('Command line option {} not supported. '
-                           'Please specify parameters in YAML file.'.format(
-                               'horizon'))
+    unknown_args = [arg for arg in args if arg.startswith("--")]
+    if any("horizon" in arg for arg in unknown_args):
+        raise RuntimeError(
+            "Command line option {} not supported."
```
Comment (ajoubertza) [inline astrokat/observe_main.py]: Missing space

File: astrokat/test/testutils.py [hunk 39]
```diff
@@ -21,28 +23,35 @@ def yaml_path(file_path):
...
         return yaml["durations"]["start_time"]
 
 
 def execute_observe_main(file_name):
-    """ Convenient method to run observer_main with correct parameters
+    """Run observer_main with correct parameters.
+
+    Parameters
+    ----------
+        file_name: relative path to yaml file
```
Comment (ajoubertza) [inline astrokat/test/testutils.py]:
```
file_name: str
    relative path to yaml file
```

File: scripts/astrokat-cals.py [hunk 40]
```diff
@@ -397,16 +391,28 @@ def table_line(datetime,  # ephem.Date object
...
-    @param notes: [optional] user provided extra information
-
-    @return: <name> <risetime UTC> <settime UTC> <Separation> <Notes>
+    Parameters
+    ----------
+        datetime: date and time
+            ephem.Date object for calculating target information
+        horizon: float
+            minimum pointing angle in degrees
+        target: str
```
Comment (ajoubertza) [inline scripts/astrokat-cals.py]: Order of params in docstring doesn't match actual params for the function.  Fix docstring.

File: scripts/astrokat-cals.py [hunk 41]
```diff
@@ -449,13 +455,19 @@ def obs_table(ref_antenna,
...
+    Parameters
+    ----------
+        ref_antenna:
+            reference location for pointing calculation as katpoint Antenna object
+        catalogue:
+            catalogue of targets as katpoint.Catalogue object
+        ref_tgt_list:
+            reference targets for calibrator selection [optional]
+        solar_sep:
+            minimum solar separation angle [optional]
```
Comment (ajoubertza) [inline scripts/astrokat-cals.py]: `lst` param missing from docstring.

**Comment by ajoubertza** [comment]
One thing that can still be added, is a copy of this file https://github.com/ska-sa/katlogger/blob/master/pyproject.toml - we want all the repo that get reformatted using [black](https://github.com/psf/black) to have this configuration.
